runtime/migrate_domain_registers.go

```diff
@@ -0,0 +1,200 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package runtime
+
+import (
+	"github.com/onflow/atree"
+
+	"github.com/onflow/cadence/common"
+	"github.com/onflow/cadence/common/orderedmap"
+	"github.com/onflow/cadence/errors"
+	"github.com/onflow/cadence/interpreter"
+)
+
+type GetDomainStorageMapFunc func(
+	ledger atree.Ledger,
+	storage atree.SlabStorage,
+	address common.Address,
+	domain string,
+) (*interpreter.DomainStorageMap, error)
+
+type DomainRegisterMigration struct {
+	ledger              atree.Ledger
+	storage             atree.SlabStorage
+	inter               *interpreter.Interpreter
+	memoryGauge         common.MemoryGauge
+	getDomainStorageMap GetDomainStorageMapFunc
+}
+
+func NewDomainRegisterMigration(
+	ledger atree.Ledger,
+	storage atree.SlabStorage,
+	inter *interpreter.Interpreter,
+	memoryGauge common.MemoryGauge,
+) *DomainRegisterMigration {
+	return &DomainRegisterMigration{
+		ledger:              ledger,
+		storage:             storage,
+		inter:               inter,
+		memoryGauge:         memoryGauge,
+		getDomainStorageMap: getDomainStorageMapFromLegacyDomainRegister,
+	}
+}
+
+// SetGetDomainStorageMapFunc allows user to provide custom GetDomainStorageMap function.
+func (m *DomainRegisterMigration) SetGetDomainStorageMapFunc(
+	getDomainStorageMapFunc GetDomainStorageMapFunc,
+) {
+	m.getDomainStorageMap = getDomainStorageMapFunc
+}
+
+// MigrateAccounts migrates given accounts.
+func (m *DomainRegisterMigration) MigrateAccounts(
+	accounts *orderedmap.OrderedMap[common.Address, struct{}],
+	pred func(common.Address) bool,
+) (
+	*orderedmap.OrderedMap[common.Address, *interpreter.AccountStorageMap],
+	error,
+) {
+	if accounts == nil || accounts.Len() == 0 {
+		return nil, nil
+	}
+
+	var migratedAccounts *orderedmap.OrderedMap[common.Address, *interpreter.AccountStorageMap]
+
+	for pair := accounts.Oldest(); pair != nil; pair = pair.Next() {
+		address := pair.Key
+
+		if !pred(address) {
+			continue
+		}
+
+		migrated, err := isMigrated(m.ledger, address)
+		if err != nil {
+			return nil, err
+		}
+		if migrated {
+			continue
+		}
+
+		accountStorageMap, err := m.MigrateAccount(address)
+		if err != nil {
+			return nil, err
+		}
+
+		if accountStorageMap == nil {
+			continue
+		}
+
+		if migratedAccounts == nil {
+			migratedAccounts = &orderedmap.OrderedMap[common.Address, *interpreter.AccountStorageMap]{}
+		}
+		migratedAccounts.Set(address, accountStorageMap)
+	}
+
+	return migratedAccounts, nil
+}
+
+func (m *DomainRegisterMigration) MigrateAccount(
```

Comment:
We might want to just expose a function that can migrate a single account, and then discuss how to use it.

Currently, at the end of a transaction, all accounts that had writes are migrated, which might take a while. 

Instead, we might want to e.g. only migrate the accounts of the transaction's authorizers.